CRM/Fieldmetadata/Normalizer.php

```diff
@@ -79,6 +80,92 @@ function setVisibilityForFields(&$fields) {
     }
   }
 
+  /**
+   * This will go through and add minDate and maxDate keys
+   * to each date field found. It will work for both custom
+   * fields and core fields.
+   *
+   * @param array $fields
+   * @return void
+   */
+  function fetchAndNormalizeDates(&$fields) {
+   
+    static $contactTypes = [];
+    static $dates = [];
+    static $formats = [];
+
+    $dateWidgets = array('Date', 'DateTime', 'Select Date', 'crm-ui-datepicker');
+    $thisYear = date('Y');
+
+    foreach ($fields as &$field) {
+      // skip non-date fields
+      if (!in_array($field['widget'], $dateWidgets)) {
+        continue;
+      }
+
+      // only do the lookup once
+      if (empty($dates[$field['name']])) {
+
+        // custom fields are little bit easier
+        if (strpos($field['name'], 'custom_') === 0) {
+          $customField = civicrm_api3('CustomField', 'getsingle', array(
+            'id' => substr($field['name'], 7),
+            'return' => 'start_date_years, end_date_years',
+          ));
+          $start = $customField['start_date_years'];
+          $end = $customField['end_date_years'];
+        }
+        // core fields
+        else {
+          // get contact types, i.e. potential entity name
+          if (empty($contactTypes)) {
+            $api = civicrm_api3('ContactType', 'get', array(
+              'sequential' => 1,
+              'is_active' => 1,
+            ));
+            foreach ($api['values'] as $contactType) {
+              $contactTypes[] = $contactType['name'];
+            }
+          }
+          
+          // look up the field first
+          $entity = in_array($field['entity'], $contactTypes) ? 'Contact' : $field['entity'];
```

Comment:
Not sure I understand this: If the entity is a contact type, then we use "Contact" as the entity, otherwise, use the passed entity?
If so, then I think we should have a comment explaining that we are normalizing the entity-string and why we need to normalize it. That we are looking up a field (and first) is pretty self-evident. Comments should say why, not what, or else they are asking to become code-rot.